core/types/subtyping.cc

```diff
@@ -1472,8 +1474,29 @@ bool Types::isSubTypeUnderConstraint(const GlobalState &gs, TypeConstraint &cons
     }
 
     if (auto *a2 = cast_type<AndType>(t2)) { // 2, 5
-        return Types::isSubTypeUnderConstraint(gs, constr, t1, a2->left, mode, errorDetailsCollector) &&
-               Types::isSubTypeUnderConstraint(gs, constr, t1, a2->right, mode, errorDetailsCollector);
+        auto subCollectorLeft = errorDetailsCollector.newCollector();
+        auto isSubTypeOfLeft = Types::isSubTypeUnderConstraint(gs, constr, t1, a2->left, mode, subCollectorLeft);
+        if (!isSubTypeOfLeft) {
+            if constexpr (shouldAddErrorDetails) {
+                auto message = ErrorColors::format("`{}` is not a subtype of `{}` (the left side of `{}`)", t1.show(gs),
+                                                   a2->left.show(gs), t2.show(gs));
```

Comment:
What do you think about just saying "(the left side of the T.all)" instead of reprinting the whole type?

I'm worried that for deeply nested types, having the parent type be reprinted is going to make it more noisy and thus hard to see what a given line is uniquely saying.

For example:

```
intersection.rb:27: Expected `T.all(M1, M2, M3, M4)` but found `C1` for argument `x` https://srb.help/7002
...
  Detailed explanation:
    `C1` is not a subtype of `T.all(M2, M3, M4)` (the right side of `T.all(M1, M2, M3, M4)`)
      `C1` is not a subtype of `T.all(M3, M4)` (the right side of `T.all(M2, M3, M4)`)
        `C1` is not a subtype of `M4` (the right side of `T.all(M3, M4)`)
```

versus

```
intersection.rb:27: Expected `T.all(M1, M2, M3, M4)` but found `C1` for argument `x` https://srb.help/7002
...
  Detailed explanation:
    `C1` is not a subtype of `T.all(M2, M3, M4)` (the right side of the T.all)
      `C1` is not a subtype of `T.all(M3, M4)` (the right side of the T.all)
        `C1` is not a subtype of `M4` (the right side of the T.all)
```

(and then imagine that instead of short, two-letter names, they're 50+ character long class names or other compound types)